### PR TITLE
Make SassCompileTask cacheable

### PR DESCRIPTION
**Pull request by britter**
- Add `@CacheableTask` annotation
- Add missing input and output declarations
- Move fallback for outDir to getOutDir()

### CODE REVIEW

**Comment by britter** [comment]
@kravemir any chance to get this merged? We're currently maintaining a fork of this plugin to get task caching.

**Comment by kravemir** [comment]
Hello @britter, thank you for contribution :slightly_smiling_face: 

> @kravemir any chance to get this merged? We're currently maintaining a fork of this plugin to get task caching.

Sorry for delayed response. I wasn't at home for the whole week.

I'll take a better look at PR in following days, and read gradle docs to understand changes. For quicker processing, I'll be glad, if you could answer questions below.

And, yes, it will get merged when PR is finished.

> * Add `@CacheableTask` annotation

Can you quickly describe, how does this change behaviour of the `AbstractSassCompileTask`?

> * Add missing input and output declarations

Thanks for fixing those annotations. How `@Internal` does change behaviour?

> * Move fallback for outDir to getOutDir()

Yes, this makes sense. However, seeing that my code as now. I wouldn't make getter to have side-effects. And, I would probably use `getEffectiveOutDir()` instead of side-effect-non-free getter.

Because, with getters with side-effects, it's impossible to do: `if ( obj.getAbc() == null ) obj.setAbc(initAbc()); `

**Comment by britter** [comment]
Hello @kravemir 

> read gradle docs to understand changes

You can start here: https://docs.gradle.org/current/userguide/build_cache.html

> Can you quickly describe, how does this change behaviour of the `AbstractSassCompileTask`?

Actually I'm not quite sure the abstract task needs the `@Cacheable` annotation at all. @breskeby can you comment?

> How @Internal does change behaviour?

Gradle scans all properties of cacheable tasks and fails if there are undeclared input or outputs (= properties that are not annotated). Some tasks may have properties that are just used to compute other properties. These properties need to be declared as [`Internal`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Internal.html) so gradle ignores them.
For example `getIncludes` and `getExcludes` are used to compute `getSassFiles`, so they do not need to be tracked as separate inputs.

> Yes, this makes sense. However, seeing that my code as now. I wouldn't make getter to have side-effects. And, I would probably use `getEffectiveOutDir()` instead of side-effect-non-free getter.

Okay, so you want this refactoring as part of this PR? I can add it then.

**Comment by kravemir** [comment]
> > read gradle docs to understand changes
> 
> You can start here: https://docs.gradle.org/current/userguide/build_cache.html

Thanks for link. I've quickly read it, and it looks like a great feature of Gradle, especially for CI/CD. 

> > Can you quickly describe, how does this change behaviour of the `AbstractSassCompileTask`?
> 
> Actually I'm not quite sure the abstract task needs the `@Cacheable` annotation at all. @breskeby can you comment?
> 
> > How @Internal does change behaviour?
> 
> Gradle scans all properties of cacheable tasks and fails if there are undeclared input or outputs (= properties that are not annotated). Some tasks may have properties that are just used to compute other properties. These properties need to be declared as [`Internal`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Internal.html) so gradle ignores them.

Thanks for explanation.

> > Yes, this makes sense. However, seeing that my code as now. I wouldn't make getter to have side-effects. And, I would probably use `getEffectiveOutDir()` instead of side-effect-non-free getter.
> 
> Okay, so you want this refactoring as part of this PR? I can add it then.

Refactoring would be fine as follow-up PR. Or, I can also do it myself, later.

File: plugin/src/main/java/org/kravemir/gradle/sass/AbstractSassCompileTask.java [hunk 1]
```diff
@@ -36,7 +38,7 @@ FileCollection getSassFiles() {
         return fileTree;
     }
 
-    @InputFiles
+    @Internal
```
Comment (kravemir) [inline plugin/src/main/java/org/kravemir/gradle/sass/AbstractSassCompileTask.java:41]: I don't think `getInputFiles` is necessary anymore. As whole `getSrcDir` is now `@InputDirectory`, then this method is not needed anymore.

It was there previously to make task rebuild CSS, if some `@import ...` file changes.

**Review by kravemir — APPROVED** [review]
However, I'm fine with yours changes. Approving, and will merge.

Thanks for contribution.

**Comment by kravemir** [comment]
@britter I've released a new version (1.2.4) containing this PR, and it's available in repositories:

* gradle plugins: https://plugins.gradle.org/plugin/org.kravemir.gradle.sass/1.2.4
* maven central: https://search.maven.org/search?q=org.kravemir.gradle.sass

**Comment by britter** [comment]
Hey @kravemir,

I was on vacation last week. Thank you for pushing out a new release. I will give it a try later today!

**Comment by britter** [comment]
We have upgraded to the new release and removed our fork. Everything works as expected. Thank you again!

**Comment by kravemir** [comment]
> We have upgraded to the new release and removed our fork. Everything works as expected. Thank you again!

Good to hear that, and you're welcome! I'm glad for yours contribution :slightly_smiling_face:

**Pull request merged**